netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -423,7 +841,478 @@ net_ebpf_extension_sock_addr_authorize_connection_classify(
+                INETADDR_SET_PORT((PSOCKADDR)&connect_request->remoteAddressAndPort, redirected_context->user_port);
+            }
+            if (address_changed) {
+                uint8_t* address;
+                if (redirected_context->family == AF_INET) {
+                    address = (uint8_t*)&redirected_context->user_ip4;
+                } else {
+                    address = (uint8_t*)&(redirected_context->user_ip6[0]);
+                }
+                INETADDR_SET_ADDRESS((PSOCKADDR)&connect_request->remoteAddressAndPort, address);
+            }
+
+            connect_request->localRedirectTargetPID = TARGET_PROCESS_ID;
+            connect_request->localRedirectHandle = redirect_handle;
+        } else {
+            InterlockedIncrement(&_net_ebpf_ext_statistics.permit_connection_count);
+        }
+    } else {
+        InterlockedIncrement(&_net_ebpf_ext_statistics.block_connection_count);
+    }
+
 Exit:
+    if (commit_layer_data) {
+        FwpsApplyModifiedLayerData(classify_handle, connect_request, 0);
+    }
+
+    NET_EBPF_EXT_RETURN_NTSTATUS(status);
+}
+
+/*
+ * Default action is BLOCK. If this callout is being invoked, it means at least one
+ * eBPF program is attached. Hence no connection should be allowed unless allowed by
+ * the eBPF program.
+ *
+ * If the eBPF program verdict is BLOCK, no context is created and AUTH classify will
+ * block the connection by default.
+ */
+void
+net_ebpf_extension_sock_addr_redirect_connection_classify(
+    _In_ const FWPS_INCOMING_VALUES* incoming_fixed_values,
+    _In_ const FWPS_INCOMING_METADATA_VALUES* incoming_metadata_values,
+    _Inout_opt_ void* layer_data,
+    _In_opt_ const void* classify_context,
+    _In_ const FWPS_FILTER* filter,
+    uint64_t flow_context,
+    _Inout_ FWPS_CLASSIFY_OUT* classify_output)
+{
+    uint32_t verdict;
+    NTSTATUS status = STATUS_SUCCESS;
+    net_ebpf_extension_sock_addr_wfp_filter_context_t* filter_context = NULL;
+    net_ebpf_extension_hook_client_t* attached_client = NULL;
+    bpf_sock_addr_t* sock_addr_ctx = NULL;
+    bpf_sock_addr_t* sock_addr_ctx_original = NULL;
+    uint32_t compartment_id = UNSPECIFIED_COMPARTMENT_ID;
+    FWPS_CONNECTION_REDIRECT_STATE redirect_state;
+    HANDLE redirect_handle;
+    uint64_t classify_handle = 0;
+    net_ebpf_extension_connection_context_t* connection_context_original = NULL;
+    net_ebpf_extension_connection_context_t* connection_context_redirected = NULL;
+    bool redirected = false;
+    FWP_ACTION_TYPE action = FWP_ACTION_BLOCK;
+    bool classify_handle_acquired = false;
+    bool v4_mapped = false;
+    bool is_original_connection;
+
+    UNREFERENCED_PARAMETER(layer_data);
+    UNREFERENCED_PARAMETER(flow_context);
+
+    if ((classify_output->rights & FWPS_RIGHT_ACTION_WRITE) == 0) {
+        // Do not modify anything and bail.
+        return;
+    }
+
+    sock_addr_ctx = (bpf_sock_addr_t*)ExAllocatePoolUninitialized(
```

Comment:
The sock_addr_ctx structs can be on stack instead of heap. Did you use heap because you think the stack will grow too large due to this?

Reply:
Yes I got analysis failure that I this function is using more than 1024 bytes on stack. So I had to switch to allocate from heap.